drivers/lora/sx1276.c

```diff
@@ -0,0 +1,717 @@
+	if (ret < 0) {
+		LOG_ERR("Unable to read OPMODE");
+		return -EIO;
+	}
+
+	regval &= ~SX1276_OPMODE_MODE_MASK;
+	regval |= SX1276_OPMODE_MODE_STDBY;
+	ret = sx1276_write(dev, SX1276_REG_OPMODE, regval);
+	if (ret < 0) {
+		LOG_ERR("Unable to write OPMODE");
+		return -EIO;
+	}
+
+	k_sleep(200);
+
+	/* Set frequency */
+	freq_rf = config->frequency;
+	freq_rf *= (1 << 19);
+	freq_rf /= CLOCK_FREQ;
+
+	ret = sx1276_write(dev, SX1276_REG_FRF_MSB, (freq_rf >> 16 & 0xFF));
+	if (!ret)
+		ret = sx1276_write(dev, SX1276_REG_FRF_MID,
+				   (freq_rf >> 8 & 0xFF));
+	if (!ret)
+		ret = sx1276_write(dev, SX1276_REG_FRF_LSB, (freq_rf & 0xFF));
+	if (ret) {
+		LOG_ERR("Unable to write RF carrier frequency");
+		return -EIO;
+	}
+
+	/* Set output power */
+	ret = sx1276_read(dev, SX1276_REG_PA_CONFIG, &pa_config, 1);
+	if (ret < 0) {
+		LOG_ERR("Unable to read version PA config");
+		return -EIO;
+	}
+
+	pa_config &= ~GENMASK(3, 0);
+#if defined CONFIG_PA_RFO_PIN
+	pa_config &= ~SX1276_PA_CONFIG_PA_BOOST;
+	if (config->tx_power < sx1276_power[SX1276_PA_RFO].min)
+		config->tx_power = sx1276_power[SX1276_PA_RFO].min;
+	if (config->tx_power > sx1276_power[SX1276_PA_RFO].max)
+		config->tx_power = sx1276_power[SX1276_PA_RFO].max;
+#else
+	pa_config |= SX1276_PA_CONFIG_PA_BOOST;
+	if (config->tx_power > sx1276_power[SX1276_PA_BOOST].max - 3) {
+		if (config->tx_power > sx1276_power[SX1276_PA_BOOST].max)
+			config->tx_power = sx1276_power[SX1276_PA_BOOST].max;
+
+		config->tx_power -= 3;
+		ret = sx1276_write(dev, SX1276_REG_PA_DAC, 0x87);
+		if (ret < 0) {
+			LOG_ERR("Unable to write PADAC");
+			return -EIO;
+		}
+	} else {
+		if (config->tx_power < sx1276_power[SX1276_PA_BOOST].min)
+			config->tx_power = sx1276_power[SX1276_PA_BOOST].min;
+
+		ret = sx1276_write(dev, SX1276_REG_PA_DAC, 0x84);
+		if (ret < 0) {
+			LOG_ERR("Unable to write PADAC");
+			return -EIO;
+		}
+	}
+#endif
+	pa_config |= (config->tx_power & 0x0F);
+	ret = sx1276_write(dev, SX1276_REG_PA_CONFIG, pa_config);
+	if (ret < 0) {
+		LOG_ERR("Unable to write PA config");
+		return -EIO;
+	}
+
+	ret = sx1276_write(dev, SX1276_REG_PREAMBLE_MSB,
+			   (config->preamble_len >> 8) & 0xFF);
+	if (ret < 0) {
+		LOG_ERR("Unable to write Preamble MSB");
+		return -EIO;
+	}
+
+	ret = sx1276_write(dev, SX1276_REG_PREAMBLE_LSB,
+			   config->preamble_len & 0xFF);
+	if (ret < 0) {
+		LOG_ERR("Unable to write Preamble LSB");
+		return -EIO;
+	}
+
+	switch (config->bandwidth) {
+	case BW_125_KHZ:
+	case BW_250_KHZ:
+	case BW_500_KHZ:
+		break;
+	default:
+		LOG_ERR("Bandwidth not supported");
+		return -EINVAL;
+	}
+
+	/* Set bandwidth */
+	ret = sx1276_read(dev, SX1276_REG_MODEM_CONFIG1, &regval, 1);
+	if (ret < 0) {
+		LOG_ERR("Unable to read Modem Config1");
+		return -EIO;
+	}
+
+	regval &= ~SX1276_MODEM_CONFIG1_BW_MASK;
+	regval &= ~SX1276_MODEM_CONFIG1_CR_MASK;
+	regval &= ~SX1276_MODEM_CONFIG1_IMP_HDR;
+	regval |= ((config->bandwidth << 4) | (config->coding_rate << 1));
```

Comment:
Can you use something like SX1276_MODEM_CONFIG1_BANDWITH_POS instead of 4 here?
Same for the rate.